File: source/kernel/fs/mbr.zig
// Copyright (c) 2025 Mateusz Stadnik
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

const std = @import("std");

pub const MBRPartitionEntry = packed struct {
    boot_indicator: u8,
    start_chs: u24,
    partition_type: u8,
    end_chs: u24,
    start_lba: u32,
    size_in_sectors: u32,
};

pub const MBR = struct {
    partitions: [4]MBRPartitionEntry,
    _signature: u16,

    pub fn create(buffer: []const u8) MBR {
        var mbr: MBR = undefined;
        for (0..4) |i| {
            mbr.partitions[i] = std.mem.bytesToValue(MBRPartitionEntry, &buffer[446 + i * @sizeOf(MBRPartitionEntry)]);
        }
        mbr._signature = std.mem.bytesToValue(u16, buffer[510..512]);
        return mbr;
    }

    pub fn is_valid(self: MBR) bool {
        return self._signature == 0xAA55;
    }
};

test "MBRPartitionEntry.ShouldHaveCorrectSize" {
    try std.testing.expectEqual(@as(usize, 16), @sizeOf(MBRPartitionEntry));
}

test "MBR.Create.ShouldParseValidMBR" {
    var buffer: [512]u8 = [_]u8{0} ** 512;
    
    // Set up a valid MBR signature
    buffer[510] = 0x55;
    buffer[511] = 0xAA;
    
    // Set up first partition entry at offset 446
    buffer[446] = 0x80; // boot indicator
    buffer[447] = 0x01; // start CHS byte 1
    buffer[448] = 0x02; // start CHS byte 2
    buffer[449] = 0x03; // start CHS byte 3
    buffer[450] = 0x83; // partition type (Linux)
    buffer[451] = 0x04; // end CHS byte 1
    buffer[452] = 0x05; // end CHS byte 2
    buffer[453] = 0x06; // end CHS byte 3
    // start_lba = 0x00001000 (little endian)
    buffer[454] = 0x00;
    buffer[455] = 0x10;
    buffer[456] = 0x00;
    buffer[457] = 0x00;
    // size_in_sectors = 0x00002000 (little endian)
    buffer[458] = 0x00;
    buffer[459] = 0x20;
    buffer[460] = 0x00;
    buffer[461] = 0x00;
    
    const mbr = MBR.create(&buffer);
    
    try std.testing.expect(mbr.is_valid());
    try std.testing.expectEqual(@as(u8, 0x80), mbr.partitions[0].boot_indicator);
    try std.testing.expectEqual(@as(u8, 0x83), mbr.partitions[0].partition_type);
    try std.testing.expectEqual(@as(u32, 0x00001000), mbr.partitions[0].start_lba);
    try std.testing.expectEqual(@as(u32, 0x00002000), mbr.partitions[0].size_in_sectors);
}

test "MBR.Create.ShouldParseInvalidSignature" {
    var buffer: [512]u8 = [_]u8{0} ** 512;
    
    // Set up an invalid MBR signature
    buffer[510] = 0x00;
    buffer[511] = 0x00;
    
    const mbr = MBR.create(&buffer);
    
    try std.testing.expect(!mbr.is_valid());
}

test "MBR.Create.ShouldParseAllFourPartitions" {
    var buffer: [512]u8 = [_]u8{0} ** 512;
    
    // Set up valid signature
    buffer[510] = 0x55;
    buffer[511] = 0xAA;
    
    // Set up all four partition entries
    for (0..4) |i| {
        const offset = 446 + i * 16;
        buffer[offset] = @intCast(i); // boot indicator
        buffer[offset + 4] = @intCast(0x0C + i); // partition type
        // start_lba
        buffer[offset + 8] = @intCast((i + 1) * 0x10);
        buffer[offset + 9] = 0x00;
        buffer[offset + 10] = 0x00;
        buffer[offset + 11] = 0x00;
        // size_in_sectors
        buffer[offset + 12] = @intCast((i + 1) * 0x20);
        buffer[offset + 13] = 0x00;
        buffer[offset + 14] = 0x00;
        buffer[offset + 15] = 0x00;
    }
    
    const mbr = MBR.create(&buffer);
    
    try std.testing.expect(mbr.is_valid());
    
    for (0..4) |i| {
        try std.testing.expectEqual(@as(u8, @intCast(i)), mbr.partitions[i].boot_indicator);
        try std.testing.expectEqual(@as(u8, @intCast(0x0C + i)), mbr.partitions[i].partition_type);
        try std.testing.expectEqual(@as(u32, @intCast((i + 1) * 0x10)), mbr.partitions[i].start_lba);
        try std.testing.expectEqual(@as(u32, @intCast((i + 1) * 0x20)), mbr.partitions[i].size_in_sectors);
    }
}

test "MBR.IsValid.ShouldReturnTrueForValidSignature" {
    var buffer: [512]u8 = [_]u8{0} ** 512;
    buffer[510] = 0x55;
    buffer[511] = 0xAA;
    
    const mbr = MBR.create(&buffer);
    try std.testing.expect(mbr.is_valid());
}

test "MBR.IsValid.ShouldReturnFalseForInvalidSignature" {
    var buffer: [512]u8 = [_]u8{0} ** 512;
    buffer[510] = 0xFF;
    buffer[511] = 0xFF;
    
    const mbr = MBR.create(&buffer);
    try std.testing.expect(!mbr.is_valid());
}

test "MBR.Create.ShouldHandleEmptyPartitions" {
    var buffer: [512]u8 = [_]u8{0} ** 512;
    buffer[510] = 0x55;
    buffer[511] = 0xAA;
    
    const mbr = MBR.create(&buffer);
    
    try std.testing.expect(mbr.is_valid());
    
    // All partitions should be zeroed
    for (mbr.partitions) |partition| {
        try std.testing.expectEqual(@as(u8, 0), partition.boot_indicator);
        try std.testing.expectEqual(@as(u8, 0), partition.partition_type);
        try std.testing.expectEqual(@as(u32, 0), partition.start_lba);
        try std.testing.expectEqual(@as(u32, 0), partition.size_in_sectors);
    }
}

test "MBR.Create.ShouldParseFAT32Partition" {
    var buffer: [512]u8 = [_]u8{0} ** 512;
    
    buffer[510] = 0x55;
    buffer[511] = 0xAA;
    
    // FAT32 partition
    buffer[446] = 0x00; // not bootable
    buffer[450] = 0x0C; // FAT32 with LBA
    buffer[454] = 0x00; // start_lba = 0x00000800
    buffer[455] = 0x08;
    buffer[456] = 0x00;
    buffer[457] = 0x00;
    buffer[458] = 0x00; // size = 0x00100000
    buffer[459] = 0x00;
    buffer[460] = 0x10;
    buffer[461] = 0x00;
    
    const mbr = MBR.create(&buffer);
    
    try std.testing.expect(mbr.is_valid());
    try std.testing.expectEqual(@as(u8, 0x00), mbr.partitions[0].boot_indicator);
    try std.testing.expectEqual(@as(u8, 0x0C), mbr.partitions[0].partition_type);
    try std.testing.expectEqual(@as(u32, 0x00000800), mbr.partitions[0].start_lba);
    try std.testing.expectEqual(@as(u32, 0x00100000), mbr.partitions[0].size_in_sectors);
}

test "MBR.Create.ShouldParseLinuxPartition" {
    var buffer: [512]u8 = [_]u8{0} ** 512;
    
    buffer[510] = 0x55;
    buffer[511] = 0xAA;
    
    // Linux partition
    buffer[446] = 0x80; // bootable
    buffer[450] = 0x83; // Linux filesystem
    buffer[454] = 0x00; // start_lba = 0x00001000
    buffer[455] = 0x10;
    buffer[456] = 0x00;
    buffer[457] = 0x00;
    buffer[458] = 0x00; // size = 0x00200000
    buffer[459] = 0x00;
    buffer[460] = 0x20;
    buffer[461] = 0x00;
    
    const mbr = MBR.create(&buffer);
    
    try std.testing.expect(mbr.is_valid());
    try std.testing.expectEqual(@as(u8, 0x80), mbr.partitions[0].boot_indicator);
    try std.testing.expectEqual(@as(u8, 0x83), mbr.partitions[0].partition_type);
    try std.testing.expectEqual(@as(u32, 0x00001000), mbr.partitions[0].start_lba);
    try std.testing.expectEqual(@as(u32, 0x00200000), mbr.partitions[0].size_in_sectors);
}

test "MBR.Create.ShouldParseCHSValues" {
    var buffer: [512]u8 = [_]u8{0} ** 512;
    
    buffer[510] = 0x55;
    buffer[511] = 0xAA;
    
    // Set CHS values for first partition
    buffer[447] = 0x12; // start CHS byte 1
    buffer[448] = 0x34; // start CHS byte 2
    buffer[449] = 0x56; // start CHS byte 3
    buffer[451] = 0x78; // end CHS byte 1
    buffer[452] = 0x9A; // end CHS byte 2
    buffer[453] = 0xBC; // end CHS byte 3
    
    const mbr = MBR.create(&buffer);
    
    try std.testing.expect(mbr.is_valid());
    
    // Verify CHS values are packed correctly (24 bits each)
    const expected_start_chs: u24 = 0x563412; // little endian
    const expected_end_chs: u24 = 0xBC9A78; // little endian
    
    try std.testing.expectEqual(expected_start_chs, mbr.partitions[0].start_chs);
    try std.testing.expectEqual(expected_end_chs, mbr.partitions[0].end_chs);
}

test "MBR.Create.ShouldHandleMaxValues" {
    var buffer: [512]u8 = [_]u8{0} ** 512;
    
    buffer[510] = 0x55;
    buffer[511] = 0xAA;
    
    // Set maximum values for first partition
    buffer[446] = 0xFF; // boot indicator
    buffer[447] = 0xFF; // start CHS
    buffer[448] = 0xFF;
    buffer[449] = 0xFF;
    buffer[450] = 0xFF; // partition type
    buffer[451] = 0xFF; // end CHS
    buffer[452] = 0xFF;
    buffer[453] = 0xFF;
    buffer[454] = 0xFF; // start_lba
    buffer[455] = 0xFF;
    buffer[456] = 0xFF;
    buffer[457] = 0xFF;
    buffer[458] = 0xFF; // size_in_sectors
    buffer[459] = 0xFF;
    buffer[460] = 0xFF;
    buffer[461] = 0xFF;
    
    const mbr = MBR.create(&buffer);
    
    try std.testing.expect(mbr.is_valid());
    try std.testing.expectEqual(@as(u8, 0xFF), mbr.partitions[0].boot_indicator);
    try std.testing.expectEqual(@as(u24, 0xFFFFFF), mbr.partitions[0].start_chs);
    try std.testing.expectEqual(@as(u8, 0xFF), mbr.partitions[0].partition_type);
    try std.testing.expectEqual(@as(u24, 0xFFFFFF), mbr.partitions[0].end_chs);
    try std.testing.expectEqual(@as(u32, 0xFFFFFFFF), mbr.partitions[0].start_lba);
    try std.testing.expectEqual(@as(u32, 0xFFFFFFFF), mbr.partitions[0].size_in_sectors);
}

File: apps/yasdisk/mbr.c
/*
 Copyright (c) 2025 Mateusz Stadnik

 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

#include "mbr.h"

#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <string.h>
#include <sys/stat.h>
#include <unistd.h>
#include <linux/fs.h>
#include <sys/ioctl.h>

// Standard CHS geometry for large disks
#define CHS_CYLINDERS 1024
#define CHS_HEADS 255
#define CHS_SECTORS 63

void format_size(uint64_t bytes, char* buf, size_t buf_size) {
    const char* units[] = {"B", "KB", "MB", "GB", "TB"};
    int unit_idx = 0;
    double size = (double)bytes;

    while (size >= 1024.0 && unit_idx < 4) {
        size /= 1024.0;
        unit_idx++;
    }

    if (unit_idx == 0) {
        snprintf(buf, buf_size, "%.0f %s", size, units[unit_idx]);
    } else {
        snprintf(buf, buf_size, "%.2f %s", size, units[unit_idx]);
    }
}

void mbr_init(MBR* mbr) {
    memset(mbr, 0, sizeof(MBR));
    mbr->signature = MBR_SIGNATURE;
}

int mbr_is_valid(const MBR* mbr) {
    return mbr->signature == MBR_SIGNATURE;
}

void mbr_print(const MBR* mbr) {
    printf("MBR Signature: 0x%04X (%s)\n", mbr->signature,
           mbr_is_valid(mbr) ? "valid" : "INVALID");
    printf("\nPartition Table:\n");
    printf("%-4s %-10s %-10s %-12s %-12s %-8s %s\n",
           "#", "Boot", "Type", "Start", "End", "Sectors", "Size");
    printf("----------------------------------------------------------------\n");

    for (int i = 0; i < MBR_PARTITION_COUNT; i++) {
        const PartitionEntry* pe = &mbr->partitions[i];
        if (partition_is_empty(pe)) {
            printf("%-4d %-10s %-10s %-12s %-12s %-8s %s\n",
                   i + 1, "No", "Empty", "-", "-", "-", "-");
        } else {
            char size_str[16];
            uint64_t bytes = (uint64_t)pe->sector_count * MBR_SECTOR_SIZE;
            format_size(bytes, size_str, sizeof(size_str));

            printf("%-4d %-10s 0x%02X %-8s %-12u %-12u %-8u %s\n",
                   i + 1,
                   partition_is_bootable(pe) ? "Yes" : "No",
                   pe->type,
                   partition_get_type_name(pe->type),
                   pe->start_lba,
                   pe->start_lba + pe->sector_count - 1,
                   pe->sector_count,
                   size_str);
        }
    }
}

void partition_init(PartitionEntry* pe) {
    memset(pe, 0, sizeof(PartitionEntry));
}

int partition_is_empty(const PartitionEntry* pe) {
    return pe->type == PART_TYPE_EMPTY;
}

int partition_is_bootable(const PartitionEntry* pe) {
    return pe->boot_flag == 0x80;
}

void partition_set_bootable(PartitionEntry* pe, int bootable) {
    pe->boot_flag = bootable ? 0x80 : 0x00;
}

const char* partition_get_type_name(uint8_t type) {
    switch (type) {
        case PART_TYPE_EMPTY:     return "Empty";
        case PART_TYPE_FAT12:     return "FAT12";
        case PART_TYPE_FAT16_S:   return "FAT16";
        case PART_TYPE_FAT16_L:   return "FAT16";
        case PART_TYPE_NTFS:      return "NTFS/exFAT";
        case PART_TYPE_FAT32:     return "FAT32";
        case PART_TYPE_FAT32_LBA: return "FAT32 LBA";
        case PART_TYPE_FAT16_LBA: return "FAT16 LBA";
        case PART_TYPE_LINUX_SWAP: return "Linux swap";
        case PART_TYPE_LINUX:     return "Linux";
        case PART_TYPE_LINUX_LVM: return "Linux LVM";
        case PART_TYPE_FREEBSD:   return "FreeBSD";
        case PART_TYPE_OPENBSD:   return "OpenBSD";
        case PART_TYPE_NETBSD:    return "NetBSD";
        case PART_TYPE_GPT:       return "GPT";
        case PART_TYPE_EFI:       return "EFI";
        default:                  return "Unknown";
    }
}

// Standard CHS calculation:
// For a given LBA:
//   cylinder = LBA / (heads * sectors_per_track)
//   temp = LBA % (heads * sectors_per_track)
//   head = temp / sectors_per_track
//   sector = temp % sectors_per_track + 1

void lba_to_chs(uint32_t lba, uint8_t* head, uint8_t* sector, uint8_t* cylinder,
                uint32_t cylinders, uint32_t heads, uint32_t sectors_per_track) {
    (void)cylinders; // Not used in calculation, for validation only

    uint32_t h = heads;
    uint32_t s = sectors_per_track;

    uint32_t cyl_val = lba / (h * s);
    uint32_t temp = lba % (h * s);
    uint32_t head_val = temp / s;
    uint32_t sector_val = (temp % s) + 1;

    // Clamp cylinder to 1023 (10 bits max for CHS)
    if (cyl_val > 1023) {
        cyl_val = 1023;
    }

    *head = (uint8_t)head_val;
    // Sector: bits 0-5 = sector number, bits 6-7 = high 2 bits of cylinder
    *sector = (uint8_t)(sector_val | ((cyl_val >> 8) << 6));
    // Cylinder: low 8 bits
    *cylinder = (uint8_t)(cyl_val & 0xFF);
}

uint32_t chs_to_lba(uint8_t head, uint8_t sector, uint8_t cylinder,
                    uint32_t heads, uint32_t sectors_per_track) {
    // Extract cylinder: lower 8 bits from cylinder byte, upper 2 from sector byte
    uint32_t cyl = cylinder | ((sector >> 6) << 8);
    uint32_t sec = sector & 0x3F;

    // LBA = (cylinder * heads + head) * sectors_per_track + (sector - 1)
    return (cyl * heads + head) * sectors_per_track + (sec - 1);
}

int disk_open(Disk* disk, const char* path) {
    memset(disk, 0, sizeof(Disk));
    strncpy(disk->device_path, path, sizeof(disk->device_path) - 1);
    disk->device_path[sizeof(disk->device_path) - 1] = '\0';
    disk->sector_size = MBR_SECTOR_SIZE;
    disk->fd = -1;

    // Get device size
    if (disk_get_size(path, &disk->device_size) != 0) {
        return -1;
    }

    // Get sector size
    disk_get_sector_size(path, &disk->sector_size);

    // Open device
    disk->fd = open(path, O_RDWR);
    if (disk->fd < 0) {
        // Try read-only
        disk->fd = open(path, O_RDONLY);
        if (disk->fd < 0) {
            return -1;
        }
    }

    // Read existing MBR
    if (disk_read_mbr(disk) != 0) {
        // Initialize empty MBR if read fails
        mbr_init(&disk->mbr);
    }

    return 0;
}

void disk_close(Disk* disk) {
    if (disk->fd >= 0) {
        close(disk->fd);
        disk->fd = -1;
    }
}

int disk_read_mbr(Disk* disk) {
    if (disk->fd < 0) {
        return -1;
    }

    if (lseek(disk->fd, 0, SEEK_SET) != 0) {
        return -1;
    }

    ssize_t n = read(disk->fd, &disk->mbr, sizeof(MBR));
    if (n != sizeof(MBR)) {
        return -1;
    }

    return mbr_is_valid(&disk->mbr) ? 0 : -1;
}

int disk_write_mbr(Disk* disk) {
    if (disk->fd < 0) {
        return -1;
    }

    // Reopen for writing if needed
    int writable_fd = disk->fd;
    struct stat st;
    if (fstat(disk->fd, &st) == 0) {
        // Check if we can write
        int flags = fcntl(disk->fd, F_GETFL);
        if ((flags & O_ACCMODE) == O_RDONLY) {
            // Reopen read-write
            writable_fd = open(disk->device_path, O_RDWR);
            if (writable_fd < 0) {
                return -1;
            }
        }
    }

    if (lseek(writable_fd, 0, SEEK_SET) != 0) {
        if (writable_fd != disk->fd) {
            close(writable_fd);
        }
        return -1;
    }

    ssize_t n = write(writable_fd, &disk->mbr, sizeof(MBR));

    if (writable_fd != disk->fd) {
        close(writable_fd);
    }

    if (n != sizeof(MBR)) {
        return -1;
    }

    // Sync to ensure write completes
    fsync(writable_fd == disk->fd ? disk->fd : writable_fd);

    disk->dirty = 0;
    return 0;
}

int disk_get_size(const char* path, uint64_t* size) {
    int fd = open(path, O_RDONLY);
    if (fd < 0) {
        return -1;
    }

    // Try BLKGETSIZE64 for block devices
    if (ioctl(fd, BLKGETSIZE64, size) == 0) {
        close(fd);
        return 0;
    }

    // Fall back to file size
    struct stat st;
    if (fstat(fd, &st) == 0) {
        *size = st.st_size;
        close(fd);
        return 0;
    }

    close(fd);
    return -1;
}

int disk_get_sector_size(const char* path, uint32_t* size) {
    int fd = open(path, O_RDONLY);
    if (fd < 0) {
        *size = MBR_SECTOR_SIZE;
        return -1;
    }

    int sect_size = 0;
    if (ioctl(fd, BLKSSZGET, &sect_size) == 0 && sect_size > 0) {
        *size = (uint32_t)sect_size;
    } else {
        *size = MBR_SECTOR_SIZE;
    }

    close(fd);
    return 0;
}

int partition_create(Disk* disk, int index, uint32_t start_lba, uint32_t sector_count, uint8_t type) {
    if (index < 0 || index >= MBR_PARTITION_COUNT) {
        return -1;
    }

    if (sector_count == 0) {
        return -1;
    }

    // Check bounds
    uint64_t end_sector = (uint64_t)start_lba + sector_count;
    uint64_t max_sectors = disk->device_size / disk->sector_size;
    if (end_sector > max_sectors) {
        return -1;
    }

    // Check for overlaps with existing partitions
    PartitionEntry temp_pe;
    partition_init(&temp_pe);
    temp_pe.type = type;
    temp_pe.start_lba = start_lba;
    temp_pe.sector_count = sector_count;

    for (int i = 0; i < MBR_PARTITION_COUNT; i++) {
        if (i != index && !partition_is_empty(&disk->mbr.partitions[i])) {
            if (partitions_overlap(&temp_pe, &disk->mbr.partitions[i])) {
                return -1;
            }
        }
    }

    PartitionEntry* pe = &disk->mbr.partitions[index];
    partition_init(pe);

    pe->boot_flag = 0x00;
    pe->type = type;
    pe->start_lba = start_lba;
    pe->sector_count = sector_count;

    // Calculate CHS (using standard 1024/255/63 geometry)
    lba_to_chs(start_lba, &pe->start_head, &pe->start_sector, &pe->start_cylinder,
               CHS_CYLINDERS, CHS_HEADS, CHS_SECTORS);

    uint32_t end_lba = start_lba + sector_count - 1;
    lba_to_chs(end_lba, &pe->end_head, &pe->end_sector, &pe->end_cylinder,
               CHS_CYLINDERS, CHS_HEADS, CHS_SECTORS);

    disk->dirty = 1;
    return 0;
}

void partition_delete(Disk* disk, int index) {
    if (index < 0 || index >= MBR_PARTITION_COUNT) {
        return;
    }
    partition_init(&disk->mbr.partitions[index]);
    disk->dirty = 1;
}

int partitions_overlap(const PartitionEntry* p1, const PartitionEntry* p2) {
    if (partition_is_empty(p1) || partition_is_empty(p2)) {
        return 0;
    }

    uint32_t p1_start = p1->start_lba;
    uint32_t p1_end = p1->start_lba + p1->sector_count - 1;
    uint32_t p2_start = p2->start_lba;
    uint32_t p2_end = p2->start_lba + p2->sector_count - 1;

    return (p1_start <= p2_end && p2_start <= p1_end);
}

int partition_validate(const Disk* disk, int index) {
    if (index < 0 || index >= MBR_PARTITION_COUNT) {
        return -1;
    }

    const PartitionEntry* pe = &disk->mbr.partitions[index];
    if (partition_is_empty(pe)) {
        return 0;
    }

    // Check bounds
    uint64_t end_sector = (uint64_t)pe->start_lba + pe->sector_count;
    uint64_t max_sectors = disk->device_size / disk->sector_size;
    if (end_sector > max_sectors) {
        return -1;
    }

    // Check for overlaps with other partitions
    for (int i = 0; i < MBR_PARTITION_COUNT; i++) {
        if (i != index && !partition_is_empty(&disk->mbr.partitions[i])) {
            if (partitions_overlap(pe, &disk->mbr.partitions[i])) {
                return -1;
            }
        }
    }

    return 0;
}

uint64_t align_sector(uint64_t sector) {
    // Align to 1 MiB boundary (2048 sectors of 512 bytes)
    const uint64_t alignment = 2048;
    return ((sector + alignment - 1) / alignment) * alignment;
}
